From c38b5c0df3f04580e453c1fe38e875dfdc2ac851 Mon Sep 17 00:00:00 2001
From: Murali Karicheri <m-karicheri2@ti.com>
Date: Thu, 1 Sep 2016 15:57:32 -0400
Subject: [PATCH 44/45] phy: keystone: fix async abort during probe on k2g soc

SerDes is part of the peripheral device. With recent change to the
keystone SerDes driver, the phy provider may get initialized
before the peripheral device to which SerDes is part of. So it
needs the associated peripheral to be powered and clocked before
initialization can happen. Without this, an async abort is seen as
below during probe due to peripheral registers not accessible when
register writes happens.

[    0.759672] ti,keystone-serdes 2320000.phy: init fw ks2_pcie_serdes.bin:
version 3.3.0.2c
[    0.768099] Unhandled fault: asynchronous external abort (0x211) at
0x00000000
[    0.775455] pgd = c0003000
[    0.778261] [00000000] *pgd=80000800004003, *pmd=00000000
[    0.783791] Internal error: : 211 [#1] PREEMPT ARM
[    0.788697] Modules linked in:
[    0.791862] CPU: 0 PID: 1 Comm: swapper Not tainted 4.4.19-02872-g3b233bc #13
[    0.799130] Hardware name: Keystone
[    0.802726] task: de050000 ti: de04e000 task.ti: de04e000
[    0.808255] PC is at kserdes_load_init_fw.constprop.3+0x194/0x1d0
[    0.814477] LR is at kserdes_load_init_fw.constprop.3+0x180/0x1d0
[    0.820697] pc : [<c02c6768>]    lr : [<c02c6754>]    psr: 60000013
[    0.820697] sp : de04fcb0  ip : 00000008  fp : de04fd24
[    0.832427] r10: de17f410  r9 : de0d3010  r8 : ffffffff
[    0.837771] r7 : f0888000  r6 : 00000000  r5 : de00bae0  r4 : de00b010
[    0.844426] r3 : f0888000  r2 : 00000000  r1 : 60000013  r0 : de1a4c80
[    0.851084] Flags: nZCv  IRQs on  FIQs on  Mode SVC_32  ISA ARM  Segment
kernel
[    0.858527] Control: 30c53c7d  Table: 00003000  DAC: fffffffd
[    0.864396] Process swapper (pid: 1, stack limit = 0xde04e208)

This patch fixes this by adding support for clock enable through
runtime PM API calls during serdes driver probe. Also add functional
clock for serdes DT node for various peripheral devices using the
serdes.

Fixes: 762fc5d069b8 ("ARM: dts: keystone: update SerDes bindings for one PHY per
lan")
Signed-off-by: Murali Karicheri <m-karicheri2@ti.com>
Signed-off-by: Jacob Stiffler <j-stiffler@ti.com>
---
 arch/arm/boot/dts/keystone-k2e-netcp.dtsi  |  6 ++++++
 arch/arm/boot/dts/keystone-k2e.dtsi        |  2 ++
 arch/arm/boot/dts/keystone-k2g-evm.dts     |  2 +-
 arch/arm/boot/dts/keystone-k2g.dtsi        | 16 ++++++++++++++--
 arch/arm/boot/dts/keystone-k2hk-netcp.dtsi |  4 ++++
 arch/arm/boot/dts/keystone-k2l-netcp.dtsi  |  4 ++++
 arch/arm/boot/dts/keystone.dtsi            |  2 ++
 drivers/phy/phy-keystone-serdes.c          | 22 +++++++++++++++++++---
 8 files changed, 52 insertions(+), 6 deletions(-)

diff --git a/arch/arm/boot/dts/keystone-k2e-netcp.dtsi b/arch/arm/boot/dts/keystone-k2e-netcp.dtsi
index 49dd715..bccbe99 100644
--- a/arch/arm/boot/dts/keystone-k2e-netcp.dtsi
+++ b/arch/arm/boot/dts/keystone-k2e-netcp.dtsi
@@ -233,6 +233,8 @@ gbe_serdes0: phy@232a000 {
 	/*rx-force-enable;*/
 	#address-cells	= <1>;
 	#size-cells	= <0>;
+	clocks = <&clkcpgmac>;
+	clock-names = "fck";
 
 	serdes0_lane0: lane@0 {
 		status		= "ok";
@@ -285,6 +287,8 @@ gbe_serdes1: phy@2324000 {
 	/*rx-force-enable;*/
 	#address-cells	= <1>;
 	#size-cells	= <0>;
+	clocks = <&clkcpgmac>;
+	clock-names = "fck";
 
 	serdes1_lane0: lane@0 {
 		status		= "disabled";
@@ -638,6 +642,8 @@ xgbe_serdes: phy@231e000 {
 	/*rx-force-enable;*/
 	#address-cells	= <1>;
 	#size-cells	= <0>;
+	clocks		= <&clkxge>;
+	clock-names	= "fck";
 
 	xserdes_lane0: lane@0 {
 		status		= "ok";
diff --git a/arch/arm/boot/dts/keystone-k2e.dtsi b/arch/arm/boot/dts/keystone-k2e.dtsi
index 1b1fc3f..ffea6c5 100644
--- a/arch/arm/boot/dts/keystone-k2e.dtsi
+++ b/arch/arm/boot/dts/keystone-k2e.dtsi
@@ -125,6 +125,8 @@
 			link-rate-kbps = <5000000>;
 			num-lanes = <2>;
 			status = "disabled";
+			clocks = <&clkpcie1>;
+			clock-names = "fck";
 		};
 
 		pcie1: pcie@21020000 {
diff --git a/arch/arm/boot/dts/keystone-k2g-evm.dts b/arch/arm/boot/dts/keystone-k2g-evm.dts
index eeab42f..53d326c 100644
--- a/arch/arm/boot/dts/keystone-k2g-evm.dts
+++ b/arch/arm/boot/dts/keystone-k2g-evm.dts
@@ -612,7 +612,7 @@
 	rx-num-evt = <32>;
 };
 
-&pcie0_phy {
+&pcie0_serdes {
 	status = "okay";
 };
 
diff --git a/arch/arm/boot/dts/keystone-k2g.dtsi b/arch/arm/boot/dts/keystone-k2g.dtsi
index 2052434..13b964a 100644
--- a/arch/arm/boot/dts/keystone-k2g.dtsi
+++ b/arch/arm/boot/dts/keystone-k2g.dtsi
@@ -845,13 +845,25 @@
 			status = "disabled";
 		};
 
-		pcie0_phy: phy@2320000 {
-			#phy-cells = <0>;
+		pcie0_serdes: phy@2320000 {
 			compatible = "ti,keystone-serdes-pcie";
 			reg = <0x02320000 0x4000>;
 			link-rate-kbps = <5000000>;
 			num-lanes = <1>;
 			status = "disabled";
+			#address-cells  = <1>;
+			#size-cells     = <0>;
+
+			power-domains = <&k2g_pds K2G_DEV_PCIE0>;
+			clocks = <&k2g_clks
+					K2G_DEV_PCIE0 K2G_DEV_PCIE_VBUS_CLK>;
+			clock-names = "fck";
+
+			pciserdes0_lane0: lane@0 {
+				status		= "ok";
+				#phy-cells	= <0>;
+				reg		= <0>;
+			};
 		};
 
 		pcie0: pcie@21800000 {
diff --git a/arch/arm/boot/dts/keystone-k2hk-netcp.dtsi b/arch/arm/boot/dts/keystone-k2hk-netcp.dtsi
index 1a6abeb..494ee00 100644
--- a/arch/arm/boot/dts/keystone-k2hk-netcp.dtsi
+++ b/arch/arm/boot/dts/keystone-k2hk-netcp.dtsi
@@ -266,6 +266,8 @@ gbe_serdes: phy@232a000 {
 	/*rx-force-enable;*/
 	#address-cells	= <1>;
 	#size-cells	= <0>;
+	clocks = <&clkcpgmac>;
+	clock-names = "fck";
 
 	serdes_lane0: lane@0 {
 		status		= "ok";
@@ -577,6 +579,8 @@ xgbe_serdes: phy@231e000 {
 	num-lanes		= <2>;
 	syscon-peripheral	= <&xgbe_subsys>;
 	syscon-link		= <&xgbe_pcsr>;
+	clocks		= <&clkxge>;
+	clock-names	= "fck";
 	/*rx-force-enable;*/
 	#address-cells	= <1>;
 	#size-cells	= <0>;
diff --git a/arch/arm/boot/dts/keystone-k2l-netcp.dtsi b/arch/arm/boot/dts/keystone-k2l-netcp.dtsi
index 9cc4be4..0f9f1ba 100644
--- a/arch/arm/boot/dts/keystone-k2l-netcp.dtsi
+++ b/arch/arm/boot/dts/keystone-k2l-netcp.dtsi
@@ -218,6 +218,8 @@ gbe_serdes0: phy@232a000 {
 	/*rx-force-enable;*/
 	#address-cells	= <1>;
 	#size-cells	= <0>;
+	clocks = <&clkcpgmac>;
+	clock-names = "fck";
 
 	serdes0_lane0: lane@0 {
 		status		= "ok";
@@ -248,6 +250,8 @@ gbe_serdes1: phy@2320000 {
 	/*rx-force-enable;*/
 	#address-cells	= <1>;
 	#size-cells	= <0>;
+	clocks = <&clkcpgmac>;
+	clock-names = "fck";
 
 	serdes1_lane0: lane@0 {
 		status		= "disabled";
diff --git a/arch/arm/boot/dts/keystone.dtsi b/arch/arm/boot/dts/keystone.dtsi
index 11dc370..ad538fd 100644
--- a/arch/arm/boot/dts/keystone.dtsi
+++ b/arch/arm/boot/dts/keystone.dtsi
@@ -304,6 +304,8 @@
 			link-rate-kbps = <5000000>;
 			num-lanes = <2>;
 			status = "disabled";
+			clocks = <&clkpcie>;
+			clock-names = "fck";
 		};
 
 		pcie0: pcie@21800000 {
diff --git a/drivers/phy/phy-keystone-serdes.c b/drivers/phy/phy-keystone-serdes.c
index 092f3c5..089142a 100644
--- a/drivers/phy/phy-keystone-serdes.c
+++ b/drivers/phy/phy-keystone-serdes.c
@@ -2474,21 +2474,37 @@ static int kserdes_probe(struct platform_device *pdev)
 	if (!sd)
 		return -ENOMEM;
 
+	pm_runtime_enable(dev);
+	ret = pm_runtime_get_sync(dev);
+	if (ret < 0) {
+		dev_err(dev, "pm_runtime_get_sync failed\n");
+		goto err_get_sync;
+	}
+
 	sd->dev = dev;
 	dev_set_drvdata(dev, sd);
 
 	ret = kserdes_of_parse(pdev, sd, np);
 	if (ret)
-		return ret;
+		goto err_of_parse;
 
 	phy_provider = devm_of_phy_provider_register(sd->dev, kserdes_xlate);
-	if (IS_ERR(phy_provider))
-		return PTR_ERR_OR_ZERO(phy_provider);
+	if (IS_ERR(phy_provider)) {
+		ret =  PTR_ERR_OR_ZERO(phy_provider);
+		goto err_of_parse;
+	}
 
 	kserdes_provider_init(sd);
 
 	dev_vdbg(&pdev->dev, "probed");
 	return 0;
+
+err_of_parse:
+	pm_runtime_put(dev);
+
+err_get_sync:
+	pm_runtime_disable(dev);
+	return ret;
 }
 
 static struct platform_driver kserdes_driver = {
-- 
1.9.1

